### PR TITLE
osd: fix cleanup job disk shredding

### LINKED ISSUE #15546: Fix cleanup job to correctly clean the disk when `cleanupPolicy` of `yes-really-destroy-data` is set.
**Is this a bug report or feature request?**
* Bug Report

Rook cleanup job cleans up the disk when `cleanupPolicy` of `yes-really-destroy-data` is set. 

After this ceph change, the bdev labels are copied to different location on the disk:
0GB, 1GB, 10GB, 100GB, 1000GB ... and so on depending of the size of the disk.
(before the change, the bdev labels were only stored in 0GB)

For example: In a 100GB disk (vdb), the labels are stored at 0GB, 1GB and 10GB.
```
$ lsblk | grep vdb
vdb    253:16   0  100G  0 disk


$ sudo hexdump -C -s 0 -n 22 /dev/vdb
00000000  62 6c 75 65 73 74 6f 72  65 20 62 6c 6f 63 6b 20  |bluestore block |
00000010  64 65 76 69 63 65                                 |device|
00000016

$ sudo hexdump -C -s 1073741824 -n 22 /dev/vdb
40000000  62 6c 75 65 73 74 6f 72  65 20 62 6c 6f 63 6b 20  |bluestore block |
40000010  64 65 76 69 63 65                                 |device|
40000016

$ sudo hexdump -C -s 10737418240 -n 22 /dev/vdb
280000000  62 6c 75 65 73 74 6f 72  65 20 62 6c 6f 63 6b 20  |bluestore block |
280000010  64 65 76 69 63 65                                 |device|
280000016
```

**Deviation from expected behavior:**
- Cleanup job is not able to clean the disks for reuse. 

**Expected behavior:**
- Cleanup job should be able to cleanup the disks for reuse. 

**How to reproduce it (minimal and precise):**

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of kubectl commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook kubectl Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/kubectl-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

### PR DESCRIPTION
**Pull request by puskunalis**
Fixes quick disk shredding by using `dd` to shred data at additional offsets where ceph metadata is duplicated. For full shred, the `shred` utility remains in use.

**Issue resolved by this Pull Request:**
Resolves #15546

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: .github/workflows/canary-integration-test.yml [hunk 1]
```diff
@@ -770,6 +770,121 @@ jobs:
...
+      - name: wait for prepare pod
+        run: tests/scripts/github-action-helper.sh wait_for_prepare_pod 3
+
+      - name: wait for ceph to be ready
+        run: tests/scripts/github-action-helper.sh wait_for_ceph_to_be_ready osd 3
+
+      - name: check-ownerreferences
+        run: tests/scripts/github-action-helper.sh check_ownerreferences
+
+      - name: teardown cluster with cleanup policy
```
Comment (sp98) [inline .github/workflows/canary-integration-test.yml:832]: Instead of a new CI to test the cleanup, I would suggest integrating this test in one of the existing CIs.

**Comment by sp98** [comment]
is this ready for review @puskunalis ?

**Comment by puskunalis** [comment]
Not yet @sp98 I will mark it as ready and re-request your review once it is

**Comment by puskunalis** [comment]
@sp98 it’s ready for review now, but I would need some guidance on what additional tests to add - encryption-pvc cleanup has issues, others have issues starting the cleanup

**Comment by sp98** [comment]
> @sp98 it’s ready for review now, but I would need some guidance on what additional tests to add - encryption-pvc cleanup has issues, others have issues starting the cleanup

what issues are you facing with the encryption-pvc ci?  
Can you also add the cleanup job logs from your local testing. Thanks.

**Comment by puskunalis** [comment]
@sp98 everything works fine locally, but as encryption-pvc sets up different volumes (/dev/mapper/...) I think it is cleared at the wrong offsets. It only works if /dev/sd* are cleaned up. This is not really related to my change (if I understand correctly), it's just that this hasn't been tested before

**Comment by travisn** [comment]
@puskunalis How is the testing going? Hopefully we can get this merged soon, thanks!

**Comment by puskunalis** [comment]
@travisn I will try to have a look into this over the next week

**Comment by puskunalis** [comment]
@sp98 could you please have a look why encryption-pvc is failing here for an encrypted PVC drive? I'm suspecting something with the test configuration, as it is not marked as a PVC OSD: https://github.com/rook/rook/actions/runs/14979187015/job/42078958350

Also, is there a reason why `shred` was used instead of `ceph-volume lvm zap` for disk cleanup? It looks as `ceph-volume lvm zap` might be the cleaner option.

**Comment by sp98** [comment]
> Also, is there a reason why `shred` was used instead of `ceph-volume lvm zap` for disk cleanup? It looks as `ceph-volume lvm zap` might be the cleaner option.

I'm not sure of the exact reason. It was written a few years ago. If you see, there are two code paths here. One for `lvm` mode OSD and other for `raw` mode OSDs. `lvm` mode OSDs use `ceph-volume lvm zap` and `raw` mode OSDs use `shred`. 
But like you said, `ceph-volume lvm zap` can be used for both the modes. The command name is confusing and one would think it only works for lvm mode, which is not true.  I also agree on just using `ceph-volume lvm zap` for all the disk cleanups.  Few thoughts around this: 

- This would mean less complexity in the code base. 
- We will have to deprecate the [`shred` utility (`sanitizeDisks`)](https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#cleanup-policy) based API in the cephCluster.  I think that should be ok because cluster cleanup is not a very common scenario and users won't mind just cleaning up the whole disk during cleanup. 
- We will lose the advantages of `shred` utility which we are anyways going if we implement `dd` for cleanup of disks. 
- There [was an issue](https://github.com/ceph/ceph/pull/62177) with `ceph-volume lvm zap` and partitioned disks. It got fixed recently. 

@travisn @puskunalis @BlaineEXE Any further thoughts on whether we just use `ceph-volume lvm zap` in the cleanup job

**Comment by puskunalis** [comment]
We could keep using shred for full disk shredding, but ceph-volume lvm zap for quick shred (otherwise we still need to use dd instead of shred)

**Comment by travisn** [comment]
I like the idea of keeping the existing shred implementation, and also in addition to run `ceph-volume lvm zap`. 

A challenge of the implementation is that currently the cleanup job runs the rook image. To run ceph-volume, we would need to run the ceph image, so it seems we need both rook and ceph images. We've done this in the past, @sp98 let's discuss.

**Comment by puskunalis** [comment]
Is ceph-volume not already in the Room image? In this PR I currently set it to run that command and the tool is present, unless the CI setup is different somehow.

**Comment by travisn** [comment]
> Is ceph-volume not already in the Rook image? In this PR I currently set it to run that command and the tool is present, unless the CI setup is different somehow.

The rook base image may be a different version from the ceph image that was deployed in the cluster according to the cluster.yaml, so it has the risk of compatibility issues. But in this case, it's a good point to consider and it likely is not worth the extra effort of running the ceph image just for the cleanup. The cleanup would be a very targeted scenario that won't change very often, so to keep the cleanup simple, it's a good point just to run the version that's already in the image. Since Rook's base image is currently v19.2.2, the latest cleanup changes would already be there.

File: pkg/daemon/ceph/cleanup/disk.go [hunk 2]
```diff
@@ -174,25 +191,107 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
...
 	defer wg.Done()
 
-	for _, device := range []string{osdInfo.BlockPath, osdInfo.MetadataPath, osdInfo.WalPath} {
+	for _, device := range []string{osdInfo.BlockPath /*, osdInfo.MetadataPath, osdInfo.WalPath*/} {
 		if device == "" {
 			continue
 		}
 
-		output, err := s.context.Executor.ExecuteCommandWithCombinedOutput(shredUtility, s.buildShredArgs(device)...)
+		// output, err := s.context.Executor.ExecuteCommandWithCombinedOutput("ceph-volume", "lvm", "zap", "--destroy", "--osd-id", strconv.Itoa(osdInfo.ID))
```
Comment (travisn) [inline pkg/daemon/ceph/cleanup/disk.go:266]: Per discussion, let's go ahead and execute this ceph-volume command, no need for complication of a separate image.

Comment (puskunalis) [reply pkg/daemon/ceph/cleanup/disk.go:266]: Sounds good, but I would need some guidance on encrypted PVC drives as the test is currently failing, it is reporting PVC name as empty, so maybe something specific with the test segtup?

Also do we need to execute this command or would it make sense to reuse the DestroyOSD function?

Comment (travisn) [reply pkg/daemon/ceph/cleanup/disk.go:266]: Oh, I overlooked that `ceph-volume lvm zap` is already called in `DestroyOSD()` [here](https://github.com/rook/rook/pull/15666/files#diff-2db24d410e5bc812291568a7c4d775d39f7b99821ec94f132748ad9d90cd6995R286). I wonder why that isn't already wiping the disk as expected? Maybe I'm missing some difference in the args.

**Comment by puskunalis** [comment]
Hey @sp98, @travisn, thanks for your patience, this is ready for review now.

Some notes:
- `ceph-volume lvm zap` doesn't work on `/dev/mapper/set1-data-0abc1-block-dmcrypt` devices, so we need to find the backing device. The method in the `DestroyOSD` function using `/mnt` doesn't work as PVC volumes aren't mounted into the cleanup pod. The function also fails to detect them as PVCs as the necessary environment variables are missing somewhere.
- `cryptsetup` is used to discover the backing device for the `/dev/mapper/set1-data-0abc1-block-dmcrypt` device. I had no luck with `lsblk`, even after changing some flags. Let me know if this can be improved.
- This way, the quick+random shred method is no longer random and is now the same as quick+zero - might need to document this somewhere? I think there is no use case for randomly shredding only tiny parts of the data on the disk.
- The cleanup flow was only added to `pvc` and `encryption-pvc` tests - are there any other tests it would make sense to be added to?
- I assume the two failing integration tests are simply flaky?
- I might have not considered encryption on non-PVC OSDs, any pointers here?

**Comment by sp98** [comment]
> Hey @sp98, @travisn, thanks for your patience, this is ready for review now.
> 
> Some notes:
> 
> * `ceph-volume lvm zap` doesn't work on `/dev/mapper/set1-data-0abc1-block-dmcrypt` devices, so we need to find the backing device. The method in the `DestroyOSD` function using `/mnt` doesn't work as PVC volumes aren't mounted into the cleanup pod. The function also fails to detect them as PVCs as the necessary environment variables are missing somewhere.
> * `cryptsetup` is used to discover the backing device for the `/dev/mapper/set1-data-0abc1-block-dmcrypt` device. I had no luck with `lsblk`, even after changing some flags. Let me know if this can be improved.

I think that should be ok.
OSD prepare pod with `mnt/` path mounted uses [EncryptionDMName](https://github.com/rook/rook/blob/a44940d924bc4231fe4bec4bcc01752e48f7d157/pkg/daemon/ceph/osd/remove.go#L271C20-L271C36) function to get the  backing device using lsblk and then clean up. 

> * This way, the quick+random shred method is no longer random and is now the same as quick+zero - might need to document this somewhere? I think there is no use case for randomly shredding only tiny parts of the data on the disk.

Agree 

> * The cleanup flow was only added to `pvc` and `encryption-pvc` tests - are there any other tests it would make sense to be added to?
> * I assume the two failing integration tests are simply flaky?

Known issue with Keystone Auth. Not related to this PR. 

> * I might have not considered encryption on non-PVC OSDs, any pointers here?

do mean not considered for covering it in CI testing? Or not considered covering it for cleanup in the cleanup job?

File: .github/workflows/canary-integration-test.yml [hunk 3]
```diff
@@ -752,6 +752,44 @@ jobs:
       - name: check-ownerreferences
         run: tests/scripts/github-action-helper.sh check_ownerreferences
 
+      - name: teardown cluster with cleanup policy
+        run: |
+          kubectl -n rook-ceph patch cephcluster rook-ceph --type merge -p '{"spec":{"cleanupPolicy":{"confirmation":"yes-really-destroy-data"}}}'
+          kubectl -n rook-ceph delete cephcluster rook-ceph
+          kubectl -n rook-ceph logs deploy/rook-ceph-operator
+          tests/scripts/github-action-helper.sh wait_for_cleanup_pod
```
Comment (sp98) [inline .github/workflows/canary-integration-test.yml:760]: So CI is using disk partitions for OSDs. Does `ceph-volume lvm zap` work for partitioned disks?  Last time I checked it didnt' work. I had opened a upstream issue for this - https://tracker.ceph.com/issues/68977.  I think it got fixed but not sure.

File: pkg/daemon/ceph/cleanup/disk.go [hunk 4]
```diff
@@ -174,32 +174,62 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
...
+	return shredCommands
+}
+
 func (s *DiskSanitizer) executeSanitizeCommand(osdInfo oposd.OSDInfo, wg *sync.WaitGroup) {
 	// On return, notify the WaitGroup that we’re done
 	defer wg.Done()
 
+	// If the device is encrypted, get the real path and remove the dm device
+	if osdInfo.Encrypted {
+		realPath, err := osd.GetBackingDeviceForEncryptedBlock(s.context, osdInfo.BlockPath)
```
Comment (sp98) [inline pkg/daemon/ceph/cleanup/disk.go:206]: So we can't use `lsblk` for this because path `/mnt` is not mounted, correct?  There is already a method for this, called [polutateDeviceInfo](https://github.com/rook/rook/blob/cf5bbd04aea0854237a9252838d142d77a0ac7a9/pkg/clusterd/disk.go#L154) but that requires `/mnt` path to be mounted.  I think current approach is also good.

Comment (puskunalis) [reply pkg/daemon/ceph/cleanup/disk.go:206]: Yes, with `PopulateDeviceInfo` it returns something like `/dev/dm-1` as `RealPath` (`NAME`), but `PKNAME` is empty even when `--no-deps` is removed from the `PopulateDeviceInfo` `lsblk` call.

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
Comment (puskunalis) [reply .github/workflows/canary-integration-test.yml:760]: It does seem to work, I'm unable to access the tracker

File: .github/workflows/canary-integration-test.yml [hunk 5]
```diff
@@ -752,6 +752,44 @@ jobs:
...
+          kubectl -n rook-ceph delete --ignore-not-found=true -f deploy/examples/operator.yaml
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:761]: To help reduce the duplication across the tests, could you refactor at least some of this into `tests/scripts/github-action-helper.sh`?

Comment (puskunalis) [reply .github/workflows/canary-integration-test.yml:761]: Sure, will do

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
Comment (sp98) [reply .github/workflows/canary-integration-test.yml:760]: I think it was fixed along with https://github.com/ceph/ceph/pull/62177

**Comment by puskunalis** [comment]
Moved the cluster teardown logic into a new function and added some docs about quick+random cleanup being the same as quick+zero. Let me know if you need me to change anything else

File: pkg/daemon/ceph/cleanup/disk.go [hunk 6]
```diff
@@ -174,32 +174,62 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
...
+		if err != nil {
+			logger.Errorf("failed to get backing device for encrypted block %q. %v", osdInfo.BlockPath, err)
```
Comment (sp98) [inline pkg/daemon/ceph/cleanup/disk.go:208]: `realpath` won't be available if there is any error.  Should we continue after the error?

Comment (sp98) [reply pkg/daemon/ceph/cleanup/disk.go:208]: @puskunalis PR looks good to me but I just one last comment ^^

Comment (puskunalis) [reply pkg/daemon/ceph/cleanup/disk.go:208]: @sp98 I'll have a look in the next few days

Comment (puskunalis) [reply pkg/daemon/ceph/cleanup/disk.go:208]: @sp98 Apologies for the delay again, I updated it so it proceeds only if there is no error

**Comment by puskunalis** [comment]
Hey @sp98 @travisn anything else needed from my side? If you'd prefer me to, I can hand over this PR to you.

File: .github/workflows/canary-integration-test.yml [hunk 7]
```diff
@@ -752,6 +752,42 @@ jobs:
       - name: check-ownerreferences
         run: tests/scripts/github-action-helper.sh check_ownerreferences
 
+      - name: teardown cluster with cleanup policy
+        run: |
```
Comment (sp98) [inline .github/workflows/canary-integration-test.yml:756]: can we use the `tests/scripts/github-action-helper.sh delete_cluster` helper for this test as well?

File: .github/workflows/canary-integration-test.yml [hunk 8]
```diff
@@ -752,6 +752,42 @@ jobs:
...
+
+      - name: wait for ceph to be ready
+        run: tests/scripts/github-action-helper.sh wait_for_ceph_to_be_ready osd 3
+
+      - name: check-ownerreferences
+        run: tests/scripts/github-action-helper.sh check_ownerreferences
+
       - name: teardown cluster with cleanup policy
         run: |
           kubectl -n rook-ceph patch cephcluster rook-ceph --type merge -p '{"spec":{"cleanupPolicy":{"confirmation":"yes-really-destroy-data"}}}'
```
Comment (sp98) [inline .github/workflows/canary-integration-test.yml:793]: can we use the `tests/scripts/github-action-helper.sh delete_cluster` helper here as well?

**Review by sp98 — COMMENTED** [review]
@puskunalis Just a couple of questions?  Other than that the PR looks good. Approving.

File: .github/workflows/canary-integration-test.yml (hunk 7 above)
Comment (travisn) [reply .github/workflows/canary-integration-test.yml:756]: let's consider this for a follow-up PR, to unblock getting this fix merged

Comment (puskunalis) [reply .github/workflows/canary-integration-test.yml:756]: Created https://github.com/rook/rook/pull/16119
